src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/SafeBuffer.cs

```diff
@@ -293,25 +300,32 @@ public void WriteArray<T>(ulong byteOffset, T[] array, int index, int count)
             if (array.Length - index < count)
                 throw new ArgumentException(SR.Argument_InvalidOffLen);
 
+            WriteSpan(byteOffset, new ReadOnlySpan<T>(array, index, count));
+        }
+
+        [CLSCompliant(false)]
+        public void WriteSpan<T>(ulong byteOffset, ReadOnlySpan<T> data)
+            where T : struct
+        {
             if (_numBytes == Uninitialized)
                 throw NotInitialized();
 
             uint sizeofT = SizeOf<T>();
             uint alignedSizeofT = AlignedSizeOf<T>();
             byte* ptr = (byte*)handle + byteOffset;
-            SpaceCheck(ptr, checked((nuint)(alignedSizeofT * count)));
+            SpaceCheck(ptr, checked((nuint)(alignedSizeofT * data.Length)));
 
             bool mustCallRelease = false;
             try
             {
                 DangerousAddRef(ref mustCallRelease);
 
-                if (count > 0)
+                if (!data.IsEmpty)
                 {
                     {
-                        fixed (byte* pStructure = &Unsafe.As<T, byte>(ref array[index]))
+                        fixed (byte* pStructure = MemoryMarshal.AsBytes(data))
```

Comment:
`MemoryMarshal.AsBytes` converts the span size to bytes and it can throw for large buffers where it would not throw before.

This should use `Unsafe.As<T, byte>` and `MemoryMarshal.GetReference` to avoid the new error condition.